pkgs/top-level/all-packages.nix

```diff
@@ -19512,6 +19512,37 @@ with pkgs;
 
   nixopsUnstable = lowPrio (callPackage ../tools/package-management/nixops/unstable.nix { });
 
+  /*
+   * Evaluate a NixOS configuration using this evaluation of NixPkgs.
+   *
+   * With this function you can write, for example, a package that
+   * depends on a custom virtual machine image.
+   *
+   * Parameter: A module, path or list of those that represent the
+   *            configuration of the NixOS system to be constructed.
+   *
+   * Result:    An attribute set containing packages produced by this
+   *            evaluation of NixOS, such as toplevel, kernel and
+   *            initialRamdisk.
+   *            The result can be extended in the modules by defining
+   *            extra options in system.build.
+   *
+   * For more information about writing NixOS modules, see
+   * https://nixos.org/nixos/manual/index.html#sec-writing-modules
+   *
+   * If you want to call this function from a system that is not
+   * supported by NixOS, you will need to initialise NixPkgs with a
+   * custom system attribute.
+   */
+  nixosEvaluate = configuration:
+    (import (pkgs.path + "/nixos/lib/eval-config.nix") {
+      inherit (pkgs) system;
+      modules = [{ config._module.args.pkgs = self; }]
```

Comment:
nit: I understand that this is supposed to be a memory optimization, but in this particular case, this modification would have no effect because the `_module.args` is an attribute set which is merged with the `//` update operator.

Also, pinning a specific version of Nixpkgs is frequent in companies, and the intent of this line would prevent the pinning from having any effect.  I guess, what you might do is add the `nixpkgs.externalPkgs` option to `nixos/modules/misc/nixpkgs.nix` and check that the user wants to use an external Nixpkgs if an another option is true.

What this module should do instead is set `nixpkgs.externalPkgs` to `self`, and let the configuration opt-in this optimization.  Note, the reason of adding an `externalPkgs` option should also benefit NixOps, which suffer from exactly the same issue.

So, I suggest to first remove this line, and do the `externalPkgs` option and setting it here in a follow-up issue.

Reply:
Thanks for catching my mistake.

I think we disagree on the purpose of the function and both purposes are legitimate, so maybe we should have two functions? Like

### `pkgs.nixos`

 * build a NixOS from current `pkgs`
 * pinned by pinning `pkgs`
 * implemented in terms of `nixosNew` with `externalPkgs` and assertions that prevent `config.nixpkgs.*` settings because they conflict.
 * does not introduce `NIX_PATH` dependency

### `pkgs.nixosNew`

 * build a NixOS from a full-blown configuration that defines its own `config.nixpkgs.*`
 * NixOS-style pinning with `config.nixpkgs.*`
 * evaluates a new NixPkgs based on config unless `externalPkgs` is set. `externalPkgs` comes with documentation about ignoring other `config.nixpkgs.*` options.
 * does not introduce `NIX_PATH` dependency

I'll focus on `externalPkgs` before continuing on this PR though.